Check for nil provider to avoid segfault in flare

Kaderinho opened the pull request:
### What does this PR do?

We had issues where the Agent panics because a Flare Provider is `nil` and we don't check for `nil` before using the provider. This PR filters all `nil` providers in the flare component before using them.

### Motivation

Avoid panic

### Possible Drawbacks / Trade-offs

Because we're filtering `nil` component, we might miss some unexpected bugs where a provider is `nil` while it should not be. In this case, we'll be missing flare information instead of panicking.

### Describe how to test/QA your changes

Run `datadog-agent flare` and `datadog-agent flare --local` and verify that the Agent does not panic.

Kaderinho committed:
fix: check for nil provider to avoid segfault

Signed-off-by: Nicolas Guerguadj <nicolas.guerguadj@datadoghq.com>

hush-hush commented on comp/core/flare/flare.go:
```diff
@@ -132,11 +132,13 @@ func (f *flare) Create(pdata ProfileData, ipcError error) (string, error) {
-			f.log.Errorf("error calling '%s' for flare creation: %s",
-				runtime.FuncForPC(reflect.ValueOf(p).Pointer()).Name(), // reflect p.Callback function name
-				err)
+		if p != nil {
+			err = p(fb)
+			if err != nil {
+				f.log.Errorf("error calling '%s' for flare creation: %s",
+					runtime.FuncForPC(reflect.ValueOf(p).Pointer()).Name(), // reflect p.Callback function name
+					err)
+			}
```
We should filter the list in the constructor instead so the rest of the code can freely use the list without worries.

Also could you add a test for this use case ?

hush-hush committed:
Filter out nil flare callback

Component might not always return a flare callback which cause FX to
populate the list with nil value.
We now filter nil provider from the provider list of the flare.

This commit also simplify the creation of a optional collector.

GustavoCaso commented on comp/core/flare/flare.go:
```diff
@@ -56,12 +56,19 @@ func newFlare(deps dependencies) (Component, rcclient.TaskListenerProvider, erro
 		collector:             deps.Collector,
 	}
 
+	// We filder nil elements from the providers list. FX doesn't filter nil elements from groups and some
```
```suggestion
	// We filter nil elements from the providers list. FX doesn't filter nil elements from groups and some
```

GustavoCaso commented on comp/core/status/statusimpl/status.go:
```diff
@@ -290,7 +291,10 @@ func (s *statusImplementation) GetStatusBySection(section string, format string,
 		}
 	default:
-		providers := s.sortedProvidersBySection[section]
+		providers, ok := s.sortedProvidersBySection[section]
+		if !ok {
+			return nil, fmt.Errorf("unknown status section '%s'", section)
+		}
```
Should we move this check to the `newStatus` constructor and filter all nil providers?

hush-hush replied:
`section` is taken as an argument and could be anything, so we need this check here.

But we should also filter nil value of all the FX group we use. I think we could do this in another PR.

GustavoCaso replied:
> section is taken as an argument and could be anything, so we need this check here.

Make sense 😄 

>  I think we could do this in another PR.

Since we are currently doing that for the flare component. I think is best to do it now, since we are currently fixing a common pattern among components that fx groups

Kaderinho committed:
fix(linter): remove redundant import

Signed-off-by: Nicolas Guerguadj <nicolas.guerguadj@datadoghq.com>

hush-hush commented:
/merge
